python/ciqueue/pytest.py

```diff
@@ -74,6 +75,43 @@ def record(self, item):
         else:
             self.redis.hdel(self.errors_key, test_queue.key_item(item))
 
+    def mark_as_skipped(self, call, item, msg):
+        assert call.when == 'teardown'
+
+        stats = self.terminalreporter.stats
+
+        def clear_out_stats(key):
+            if key in stats:
+                new_stats = []
+                for i in stats[key]:
+                    if i.nodeid != item.nodeid:
+                        new_stats.append(i)
+                    elif self.logxml:
+                        xmlkey = 'failure' if key == 'failed' else key
+                        self.logxml.stats[xmlkey] -= 1
+                stats[key] = new_stats
+                if not stats[key]:
+                    del stats[key]
+
+        # remove the failure/error from logxml
+        if self.logxml:
+            self.logxml.node_reporters_ordered[-1].nodes = []
+
+        # the call is converted to a skip
+        call.excinfo = outcomes.skipped_excinfo(item, msg)
+
+        # clear out the stats like the test never happened
+        for key in ('passed', 'error', 'failed'):
+            clear_out_stats(key)
+
+        # rollback the testsfailed number like it never happened
+        item.session.testsfailed -= len([v for k, v in item.error_reports.items()
+                                         if not issubclass(v['excinfo'].type, outcomes.Skipped) and k != 'teardown'])
```

Comment:
Are failed teardown's counted elsewhere?

Reply:
We mark the test as skipped after setup and test, but before reporting on teardown. What we're doing here is rolling back setup and test, and replacing teardown's result with skipped. So the reason we don't have to rollback teardown is because we haven't reported on it yet. At the top of the function, we assert that this is only called on teardown.